### ISSUE TITLE
QUESTION: Is this advisable to do in a handler?

### ISSUE THREAD

**Issue by relaxdiego** [canonical/operator#144]
Given a handler for `on.config_changed`:

```python
def on_config_changed(self, event):
    unit_status = None

    while not isinstance(unit_status, ActiveStatus):
        # CODE THAT DOES STUFF & SETS unit_status HERE

        self.framework.model.unit.status = unit_status
```

Is that `while` loop advisable? Will Juju timeout the code if it ever runs too long?

**Comment by dshcherb** [comment]
In short: no, don't do that.

Status values are for operators to observe via Juju status and your charm code must not depend on those itself (only set them). If really you need some persistent state to be used across hook executions, use attributes on `StoredState`.

Hook executions have a very short lifetime: if you don't have enough information you either exit and (implicitly) wait for another event to happen or you defer an event and exit so that it is reemitted on the next hook execution.

Not having enough information in your current context is not an error condition by itself and you should not block on it in your charm code because you always work with a snapshot of data from a controller. The current state at the controller is cached by a unit agent as you access it: e.g. `relation-get` would return the latest relation data at the time of `relation-get` execution once and then will always return the same thing in the current hook execution. If this cached data is not enough to progress the charm code forward - you simply exit from the current hook execution (`return` from the event handler).

If your code polls `relation-get` or some other data from the controller via hook tools in that `while` loop, it will do so infinitely getting the same cached data from the unit agent and will never progress.

**Comment by relaxdiego** [comment]
In my case, however, I'm not reading data provided by the current context. Rather, I'm polling the underlying k8s API directly to see if the unit's associated pod is ready to serve requests (i.e. livenessProbe and readinessProbe both return 20x). So there's not chance of me viewing cached data.

The reason I'm doing that is so that my unit's status stays in "maintenance" until the pod is actually ready to serve requests.

This was me just messing around with the framework to see what works and what doesn't. I could try using `defer()` but I found that it's not always guaranteed that the event will be re-emitted. For example when it's the only charm running and there are no changes to relations. In that case, `on.config_changed` or `on.update_status` won't run and my application's status will be stuck in maintenance.

**Comment by relaxdiego** [comment]
So here's a sample handler where I tried to defer it:

```python
def on_update_status(self, event):
    msg = str(datetime.datetime.now())
    self.adapter.set_unit_status(MaintenanceStatus(msg))
    event.defer()
```

I'm just using the datetime check if it ever gets called multiple times. It doesn't and the time is stuck at the first time msg was defined:

```
Model       Controller  Cloud/Region        Version  SLA          Timestamp
prometheus  mk8s        microk8s/localhost  2.7.1    unsupported  18:26:26+08:00

App                   Version  Status  Scale  Charm                 Store  Rev  OS          Address        Notes
charm-k8s-prometheus           active      1  charm-k8s-prometheus  local    0  kubernetes  10.152.183.50

Unit                     Workload     Agent  Address    Ports     Message
charm-k8s-prometheus/0*  maintenance  idle   10.1.43.4  9090/TCP  2020-02-18 10:23:08.052926
```

Here's a more full example of the `while` loop hack. It ain't pretty yet because I just wanted to see it work for now: https://github.com/relaxdiego/charm-k8s-prometheus/blob/wait-for-pod-readiness/src/charm.py#L51

**Comment by relaxdiego** [comment]
Alright. I take that last part back about on.update_status not being reemitted. The Message does update once every 4~5 minutes:

```
2020-02-18 10:23:08.052926
2020-02-18 10:28:54.197650 <— duration=00:05:46.144724
2020-02-18 10:33:25.906317 <— duration=00:04:31.708667
2020-02-18 10:38:07.526588 <— duration=00:04:41.620271
```

NEXT QUESTION: Is there any way to make it reemitted faster than that?

**Comment by relaxdiego** [comment]
I'm probably looking at this the wrong way, am I? I suppose no charm operates in a vacuum and there's a higher probability than I originally assume there is that `on.update_status` will be reemitted sooner, particularly if a relation is created.

**Comment by jameinel** [comment]
There are ways to change the interval for update-status, but they are meant
more for Admins than for Charms. (eg, in this deployment, I want
faster/slower status updates). The average interval can be set in
model-config and Juju intentionally randomizes the time interval by +/- 20%
so that not every unit of every application wants to run the same hook at
exactly the same time.

I don't think looping on self.framework.model.unit.status would be the
right check, but looping on "is the K8s pod up and running" is actually ok.
If you take K8s out of the conversation for a moment. A simple 'postgresql'
charm could quite reasonably 'apt install postgresql' and wait until it can
tell that the postgresql app is able to respond before proceeding.

The things to consider are:

   - Only one hook fires at a time. So as long as you are blocking waiting
   for something, you can't respond to other updates in the model. if it makes
   sense to do so, then its fine. (eg, I can't respond to a relation-joined
   event to tell them who to connect to if my application isn't up and running
   yet.)
   - Status is only reported to humans (other charms don't see the status
   of their related applications). The information you get to send to other
   related units is what you put in 'relation-set'. However, the data from
   'relation-set' is not committed to the database and available to other
   units until your hook exits (because until then you may get an exception in
   your code). This is what allows you to see a 'snapshot' of the world around
   you when the hook starts, and not have to worry midway through the hook
   that something you just looked at has changed. If it *does* you'll get
   another event. (eg, you can call relation-get to find out what another unit
   is telling you. If your code calls relation-get 1s before/after the other
   charm calls relation-set, Juju will ensure you get another relation-changed
   event.)
   - So you certainly *shouldn't* ever loop waiting for the Juju data to
   change around you, as we carefully don't let you see it :)

**Comment by relaxdiego** [comment]
Thank you for the detailed and very helpful insights @jameinel! This gives me enough to move forward with my charm development.

**Comment by relaxdiego** [comment]
Just to close the loop on this question. Here's my implementation: https://github.com/relaxdiego/charm-k8s-prometheus/blob/0f60d0bb890097b7ab988b957dcf45945714861f/src/charm.py#L51

**Comment by jameinel** [comment]
Thanks for your follow up.
I do have some concerns that we are doing you a disservice if the way to
"interact with the framework" is to immediately write an adapter that moves
all of your code away from the framework. Having you add a thunk means that
what we're giving you isn't what you find useful. I'm not saying your code
isn't very helpful for you, but I'd like to understand so that I can
understand where our gaps are.

Things like:
https://github.com/relaxdiego/charm-k8s-prometheus/blob/0f60d0bb890097b7ab988b957dcf45945714861f/src/handlers.py
mean you are expressly moving away from having the state of a Class to bind
together common logic and state, in favor of doing it as free form
functions.

**Comment by relaxdiego** [comment]
Thanks for your message @jameinel. I truly appreciate you taking the time to look at my code and providing feedback given you and the rest of the team's tight schedule.

> I do have some concerns that we are doing you a disservice if the way to "interact with the framework" is to immediately write an adapter that moves all of your code away from the framework.

You are not. On the contrary, I am enjoying my experience writing charms for the Operator Framework and I commend the core team for all their effort thus far.

One of the reasons for the adapter has to do with ensuring that my charm's tests are as easy to follow as possible because, based on over 2 decades of writing unit tests, the truly hard part about automated tests is when the SUT fails one of them and one has to read through the failing test to understand the reason for the failure. From experience, it's best to have everything that one needs to understand the failing test explicitly written in said test. Specifically, one only need to read the test procedurally from top to bottom and get what it's doing by the time one has read its last line of test code. I find that if I have to jump back and forth between the test and a few helper methods then it makes the test hard to follow. Questions like:

> What does `self.do_that_thing_with_the_backend()` have to do with this test?

...or:

> I see that `self.create_mock_thingamabob(set_foo=True, bar=some_obj)` is preparing a mocked object that this test needs but which part of this mock is absolutely needed by this test and which ones are unnecessary?

...will get in the way of quickly understanding the test. With tight deadlines and a gazillion fires to fight, more often than not, the "solution" is to comment out the test just to move forward, promising oneself that one will get back to it sometime in the future. I'm going to guess I don't need to tell the parable of the "interim solution" here. :-)

By using an adapter, the unit test doesn't have to have any extraneous lines of code. Every line has a purpose in the current test and there's no line of test code that's relevant for other tests but irrelevant for the current one. Here is an example. Notice how its use of `mock_adapter` is very specific: https://github.com/relaxdiego/charm-k8s-prometheus/blob/0015fa53d1cc20be227bbc1150468f72b188e13c/test/charm_test.py#L80

Another reason for the adapter is beneficial to both the framework core team and the framework users: It insulates the core team from being locked in to a public API implementation too early in the framework's life and it protects the framework users from having to do an excruciating "seek and replace" in their code when the core team decides to change the public API. This is because the adapter decouples the frameworks public API from the framework consumers to a certain degree.

The above benefit does come with its own disadvantage if not managed early. That is, the adapter's expectation of the underlying framework might go out of sync but the tests won't show it. To counter this, the adapter's tests must be well maintained and make use of `autospec=True` and `spec_set=True` when mocking the framework. Example: https://github.com/relaxdiego/charm-k8s-prometheus/blob/master/test/adapter_test.py

---

> Things like:
https://github.com/relaxdiego/charm-k8s-prometheus/blob/0f60d0bb890097b7ab988b957dcf45945714861f/src/handlers.py
mean you are expressly moving away from having the state of a Class to bind
together common logic and state, in favor of doing it as free form
functions.

This is the "imperative shell + functional core" design pattern as described by Gary Bernhardt in his talk titled [Boundaries (length: 30:19)](https://pyvideo.org/pycon-us-2013/boundaries.html). I tend to use this architectural design pattern when coding because it allows for unit tests that reflect reality yet easy to maintain. You'll see this reflected in my Charm (imperative shell) tests: https://github.com/relaxdiego/charm-k8s-prometheus/blob/master/test/charm_test.py and my handlers (functional core) tests: https://github.com/relaxdiego/charm-k8s-prometheus/blob/master/test/handler_test.py.

Some of the benefits of this architecture are:

1. O(n) test execution time where n is the number of branches in the codebase. This is due to the fact that you are writing strictly isolated unit tests rather than integration tests pretending to be unit tests which results in permutations that grow faster than O(n). Here's the build history of one of my charms: https://travis-ci.org/relaxdiego/charm-k8s-prometheus/builds
2. Less reliance on external dependencies in one's core since those dependencies never cross the boundary between the imperative shell and the functional core. Rather, the functional core only receives values (The value is the boundary as per Gary Bernhardt's talk). This results in really simple functional core tests that are easy to reason about.
3. Just as important: the imperative shell code becomes very procedural, making it easy to follow and the functional core's code while having the tendency to create branches, are still easy to follow since it is free of terminologies (variable names, method names) that belongs in the imperative shell layer. The terminologies in the functional core is free to be strictly business logic specific.

There are many more advantages to this architectural design decision but the above should suffice for now. Happy to discuss this some more with you. Again, thank you so much for taking the time to give me your feedback and please don't hesitate to ping me for other clarifications.
